### ISSUE TITLE
Avoidable circular reference error in types

### ISSUE THREAD

**Issue by inad9300** [microsoft/TypeScript#45805]
# Bug Report

### 💻 Code

```ts
function registerEvents<T extends HTMLElement>(elem: T, events: Record<string, () => void>): void {}

const button = document.createElement('button')

const x = registerEvents(button, {
    click: () => console.log(x)
})
```

### 🙁 Actual behavior

An error is reported by the compiler: `'x' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

### 🙂 Expected behavior

No error reported, as it is trivial to infer the type of `x` (`void`) without getting into circular references. The code above should behave exactly the same as the following code:

```ts
function registerEvents<T extends HTMLElement>(elem: T, events: Record<string, () => void>): void {}

const button = document.createElement('button')

const x = registerEvents(button, {
    click: () => { console.log(x) }
})
```
[Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYHMYGcquQUQDdUwosAeAFUVQA9cwATLRACUoFkAZfAG1QC2JKAD4AFKn4CAXIkoAaGsVJZZAJVQQ4yRuRzIYYdIrEBKRAF4RiQnBiMRp2bfuIA3gF8AUF61gciABGIFBQCJaIjHAQIEKkAHQQaACGuHyCwmIA5MGhCFmmPn4BtBFomDh4RMJYYrlhYIpuXoitiBC8MBAA1rJmltZu7QhYcPzxvHDoYrTm3h6FQA)

Template parameters seem to be playing a role. The following code also poses no problem to the compiler:

```ts
function registerEvents(elem: HTMLElement, events: Record<string, () => void>): void {}

const button = document.createElement('button')

const x = registerEvents(button, {
    click: () => console.log(x)
})
```
[Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYHMYGcquQUQDdUwosAKVAG1QFsAuRACQBUBZAGXxtpKgBpEqYqSyMASqghxkAEwA8OZDDDpB5AJSIAvAD5EhODFm6NjQ8cQBvAL4AoO9LA5EAIxBQoCHYllwIILykAHQQaACGuNx0fOQA5O6eCHEaDk4uAB4+aJg4eER8FIleYIJWdoiViBBUMBAA1oyaOvrpcDTBVHDo5BmpNqlAA)

**Comment by andrewbranch** [comment]
If you search for `it does not have a type annotation and is referenced directly or indirectly in its own initializer`, they all look avoidable from a human perspective. The reality is we _do_ get into circular references, it just happens that in this case, the type of `x` in `console.log(x)` doesn’t matter. But figuring out whether or not it matters in the general case is not trivial, or at least, it’s a known limitation of our checker.

**Comment by inad9300** [comment]
I don't know about other cases, but what's remarkable about this one is that calling a function with two objects which differ only syntactically but have the exact same type produces two different results (first error-free snippet). Additionally, when changing the type of the first argument and leaving the second argument (presumably the problematic one) untouched, this also changes the behavior (see second error-free snippet).

**Comment by andrewbranch** [comment]
Those two cases only have the same type coincidentally—the blockless arrow function’s return type is determined by the type of the expression `console.log(x)`, whereas the block-bodied arrow function has the return type `void` because it lacks any `return` statement. So it’s not really just a syntactic difference; it’s a semantic difference that happens not to be relevant to the resulting type (or rather, wouldn’t be relevant if we were smart enough to analyze it without error).

**Comment by inad9300** [comment]
I don't know, you are the experts here, but from the outside it seems that determining the type of `() => console.log(x)` is something that TypeScript is more than capable of doing, knowing that the return type of `console.log` is static and explicitly stated in the program text. If the current design has limitations that make fixing this problem so difficult, please consider changing the design.
